### DISCUSSION TITLE
Using fail2ban over longer time spans?

### DISCUSSION THREAD

**Discussion by richb-hanover** [fail2ban/fail2ban#2952] [Q&A]
I am using some homegrown scripts that look through log files and update iptables rules. Here's my current process:

- I combine `/var/log/kern.log.*` files for the last *two days*
- I then grep out certain lines and count the occurrences of activity from the IP addresses
- For those IP addresses that exceed a threshold (over those two days), I create an iptables rule to drop/reject their traffic
- I currently drop/reject the address forever, but might be willing to re-enable them after a long time, say 7 or 14 days

I just realized that fail2ban uses a lot of the same machinery.

Is this something that fail2ban could accommodate? Or should I stick with my current scripts? Many thanks!

**Comment by sebres** [comment]
> I combine log files for the last two days

Fail2ban uses real-time monitoring either (but you can indeed define `findtime = 2d` to consider matches in two days window).
Note that 2 days is a long time, so options like `maxmatches` (see #2402) could be set to 0 to save memory. 

> I then grep out certain lines and count the occurrences of activity from the IP addresses

It depends on kind of activity you need to consider, fail2ban would count all matches (by regex) during `findtime` interval and if it exceeds `maxretry` will arrange a ban for such IPs. There is also bantime increment feature which would reduce maxretry (so ban earlier) and increase bantime for known as bad, recidive IPs.

> For those IP addresses that exceed a threshold ...

If it is simple threshold calculation (like fail2bans `maxretry` attempts during `findtime` interval), then you can do that with fail2ban also.
Also note https://github.com/fail2ban/fail2ban/issues/1945#issuecomment-362247222 if you're trying to monitors network bots or port scanner or crawlers.

> I currently drop/reject the address forever, but might be willing to re-enable them after a long time, say 7 or 14 days

Persistent banning is something ugly (just an unneeded last on net-subsystem etc), and is obsolete once fail2ban gets bantime increment feature in v.0.11... but if you want it nevertheless - set `bantime = -1`.
Recommended setting for long time banning would be something like that:
```ini
[jail]
# initial ban time:
bantime = 1h
# incremental banning:
bantime.increment = true
# default factor (causes increment - 1h -> 1d 2d 4d 8d 16d 32d ...):
bantime.factor = 24
# max banning time = 5 week:
bantime.maxtime = 5w
```
This will firstly ban for 1 hour (this way not so aggressive by false positives for some legitimate users doing some mistakes, but it also allows to reduce initial `findtime` and `maxretry` so can ban earlier) and then by repeated bans causes exponential growth starting from 1 day up to 5 week (restricted by `bantime.maxtime`).
If you need a "jump" 1h -\> 7d -\> 14d ... you can use 168 as `bantime.factor`.

> Is this something that fail2ban could accommodate?

What exactly?

**Reply by richb-hanover** [reply]
Thank you for this thorough response! Here's more information about my situation:

I run a netperf server that people use to test their bandwidth. I am delighted to make the server available at no cost to researchers who are studying [bufferbloat](https://www.bufferbloat.net/projects/bloat/wiki/What_can_I_do_about_Bufferbloat/) and to random people who want to decrease lag/latency in their home networks.

But the problem I face is that some people run automated speed tests every 5 minutes, 24 x 7, and the VPS that's hosting burns through 4TB/month in a few days, and my hosting company shuts me off. 

So... I need a way to block out the abusers while allowing the "good guys" through. My heuristic is that people doing research make fewer than 35 "speed tests" in two days. (Each speed test usually creates 10 TCP connections, so my threshold is 350 connections in a two-day span.) I have an iptables rule to log each speedtest connection, then I count those connections over a two-day span, and create another iptables rule to drop/reject addresses that exceed the threshold.

My "willingness to forgive/un-ban" is pretty small - if some IP address exceeds the threshold, the owner probably has set up an automated process. They're not going away. (Although I am considering throwing them into an iptables chain that only permits 2 packets per second... That'll knock their bandwidth measurement *way* down, and they'll think my test site is broken. And then they'll go away :-)

So - given that I have all this machinery substantially working, do you see any benefit to switching to fail2ban? Thank you again for the great response.

**Reply by novazur972** [reply]
> If you need a "jump" 1h -> 7d -> 14d ... you can use 168 as `bantime.factor`

I can't understand that.
1h * 168 = 7d : OK
but
7d * 168 = 14d ? or 1176d ?
Is it me who doesn't understand the principle?

**Reply by sebres** [reply]
The principle is:
- either `ban.Time * banFactor * (1<<(ban.Count if ban.Count<20 else 20))`
- or `ban.Time * banFactor * multipliers[ban.Count if ban.Count < len(multipliers) else -1]` (if `bantime.multipliers` are used)

What in normal case would mean that by factor 168 the increment will be `1h -> 14d -> 28d -> 56d ...` (factor multiplies initial `bantime`). I miscalculated here a bit: for `1h -> 7d` jump, the factor must be 84, then it'd get `1h -> 7d -> 14d -> 28d -> 56d ...`

If you need different growth, either use `bantime.multipliers` or rewrite `bantime.formula`...
For instance `ban.Time * (1<<(banFactor*(ban.Count if ban.Count<20 else 20)))` would grow drastic and effect of factor gets exponential.

PoC:
```bash
$ python3 -c 'factor = 84; print("\n".join(["%4d.) %s d" % (count, ((1*60*60 * factor * (1<<count)) / (24*60*60))) for count in range(1,5)]))'
   1.) 7.0 d
   2.) 14.0 d
   3.) 28.0 d
   4.) 56.0 d
$ python3 -c 'factor = 168; print("\n".join(["%4d.) %s d" % (count, ((1*60*60 * factor * (1<<count)) / (24*60*60))) for count in range(1,5)]))'
   1.) 14.0 d
   2.) 28.0 d
   3.) 56.0 d
   4.) 112.0 d

# factor affects exponentialy:
$ python3 -c 'factor = 3; print("\n".join(["%4d.) %s d" % (count, ((1*60*60 * (1<<(count*factor))) / (24*60*60))) for count in range(1,5)]))'
   1.) 0.3333333333333333 d
   2.) 2.6666666666666665 d
   3.) 21.333333333333332 d
   4.) 170.66666666666666 d
$ python3 -c 'factor = 4; print("\n".join(["%4d.) %s d" % (count, ((1*60*60 * (1<<(count*factor))) / (24*60*60))) for count in range(1,5)]))'
   1.) 0.6666666666666666 d
   2.) 10.666666666666666 d
   3.) 170.66666666666666 d
   4.) 2730.6666666666665 d

```

**Reply by novazur972** [reply]
> The principle is:

It's very kind of you to have explained all this to me but sorry, I know nothing about python, and above all, I understand even less now how a factor of 168 can go from 1h to 14 days.
For me, stupidly, 1hx168 = 168h = 168/24 days = 7 days.
Everything else is a pure mystery.
But don't bother any more. I didn't think you had to be a python dev to configure fail2ban. I will continue to define multipliers, it is probably the simplest at my level.

**Reply by sebres** [reply]
>  I know nothing about python, and above all, I understand even less now how a factor of 168 can go from 1h to 14 days.

But you surely understand math, don't you?

`t = 1·60·60 · f · (1<<c) / (24·60·60)`
or the same using power of base 2, because <code>(1<<c) == 2<sup>c</sup></code>:
<code>t = 1·60·60 · f · (2<sup>c</sup>) / (24·60·60)</code>

where `1·60·60` - one hour in seconds, `f` - factor, `c` - known ban counts of IP, `24·60·60` - 24 hours (1 day) in seconds

as one can see the only exponent is depending on `c` (ban-count), not on factor, which only affects (initial) bantime.

Calculated <code>t = 1·60·60 · f · (2<sup>c</sup>) / (24·60·60)</code>:
c | t for (f = 84) | t for (f = 168) | explanation
--|--|--|--
0 | 1h | 1h | <code>1h · f · 2<sup>0</sup></code>
1 | 7d | 14d | <code>1h · f · 2<sup>1</sup></code>
2 | 14d | 28d | <code>1h · f · 2<sup>2</sup></code>
3 | 28d | 56d | <code>1h · f · 2<sup>3</sup></code>
4 | 56d | 112d | <code>1h · f · 2<sup>4</sup></code>

that would change if the formula gets rewritten in the way that factor becomes an exponent, e. g.  <code>f · 2<sup>c</sup></code> => <code>2<sup>c·f</sup></code>. Then the progression would look like your initial assumption (but 84 or 168 were surely too large factors).

> Everything else is a pure mystery.

No, it is pure math... But OK...

> I didn't think you had to be a python dev to configure fail2ban. 

You don't. 
The python snippets, I provided above, can be used to estimate different factors (or formulas) or their effect on the incremented bantime. 

I like people asking something firstly and then if the answer expects to turn on the brain a little bit, abrupt it becomes a "pure mystery". Education nowadays.

**Comment by sebres** [comment]
> do you see any benefit to switching to fail2ban?

You can try it. The config for latest 0.11 will be something like this:

`fail2ban.local:`
```ini
[DEFAULT]
dbmaxmatches = 0
dbpurgeage = 20d
```
`jail.local:`
```ini
[DEFAULT]
maxmatches = 0

[sptestlimit]
logpath = /var/log/kern.log
filter =
failregex = ^\s*\S+\s+kernel:(?: +\[[^\]]+\])? Speed-Test (?:\b(?:IN=\w+|OUT=|(?:(?!OUT=|IN=)[A-Z]+=[^ \[]*)+) )*SRC=<ADDR> DST=\S+
findtime = 2d
maxretry = 350
bantime = 1d
bantime.increment = true
bantime.factor = 7
bantime.maxtime = 5w
banaction = iptables-ipset-proto6-allports
```

This configures a jail `sptestlimit` which would ban IPs generating 350 attempts in 2 days interval for 1 day, hereafter after 175 attempts for 7 days, etc.
The `failregex` matching lines like this (that fail2ban will consider as a failure):
```
Feb 28 23:45:30 host kernel: [5140091.073466] Speed-Test IN=eno1 OUT= MAC=... SRC=192.0.2.1 DST=127.0.0.1 ...
```
If your line format for speed test attempt looks different, you must adjust `failregex`.
As banning action it uses `iptables-ipset-proto6-allports` (I think ipset is better for a lot of long banned tickets), but there are other you could use (like `iptables-allports` or `iptables-multiport`) or you can write your own custom action, see [/etc/fail2ban/action.d](https://github.com/fail2ban/fail2ban/tree/0.11/config/action.d) directory for examples.

As for possible benefits, there are indeed some probably:
- fail2ban doing monitoring continuously (so this 2 day window "scrolls" quasi all the time by counting of the attempts), so it is more dynamic and you can probably decrease `findtime` together with `maxretry` to something like 5h and 35 (which will be almost the same than 350 attempts in 2 days, but the ban will occur much faster);
- one can create multiple jails (so able to have different burst/thresholds as well as different punishment scenarios, so more precise tuning is possible);
- fail2ban holds evildoers in sqlite database, so for example able to restore banned IPs after restart (and you could get some statistic from there, etc);
- due to bantime increment it is really not necessary to ban persistently (it will ban known evildoers faster and longer automatically)

**Reply by richb-hanover** [reply]
This is fabulous information! It does sound as if fail2ban could work better. 

* ipset does make sense: my long list of addresses is clearly not optimal. Using fail2ban's machinery means I don't have to debug it.
* You're right about having a shorter window. Shorter window -> tighter control on abuser
* A standard method of controlling traffic, instead of "Rich's Homegrown Scripts". (But they are organic :-)
* All-in-all, a better scheme. 

Thanks again
